chromium/background.js

```diff
@@ -193,29 +177,70 @@ function AppliedRulesets() {
 }
 
 AppliedRulesets.prototype = {
-  addRulesetToTab: function(tabId, ruleset) {
-    if (tabId in this.active_tab_rules) {
-      this.active_tab_rules[tabId][ruleset.name] = ruleset;
+  addRulesetToTab: function(tabId, type, ruleset) {
+    if (!this.active_tab_main_frames.has(tabId)) {
+      this.active_tab_main_frames.set(tabId, false);
+    }
+
+    // always show main_frame ruleset on the top
+    if (type == "main_frame") {
+      this.active_tab_main_frames.set(tabId, true);
+      this.active_tab_rules.set(tabId, [ruleset,]);
+      return ;
+    }
+
+    if (this.active_tab_rules.has(tabId)) {
+      let rulesets = this.active_tab_rules.get(tabId);
+      let insertIndex = 0;
+
+      const ruleset_name = ruleset.name.toLowerCase();
+
+      for (const item of rulesets) {
+        const item_name = item.name.toLowerCase();
+
+        if (item_name == ruleset_name) {
+          return ;
+        } else if (insertIndex == 0 && this.active_tab_main_frames.get(tabId)) {
+          insertIndex = 1;
+        } else if (item_name < ruleset_name) {
+          insertIndex++;
+        }
+      }
+      rulesets.splice(insertIndex, 0, ruleset);
     } else {
-      this.active_tab_rules[tabId] = {};
-      this.active_tab_rules[tabId][ruleset.name] = ruleset;
+      this.active_tab_rules.set(tabId, [ruleset,]);
```

Comment:
this will be necessary when there is no applicable ruleset applied to the `main_frame` but at least one ruleset applied to the page resources; this is typical when a HTTP-only site serving third-parties HTTPS content.